src/docfx/cli/Program.cs

```diff
@@ -9,23 +10,43 @@ internal class Program
     {
         internal static async Task Main(string[] args)
         {
-            var (command, docset, options) = ParseCommandLineOptions(args);
+            var (command, source, options) = ParseCommandLineOptions(args);
             var log = new ConsoleLog();
 
             switch (command)
             {
                 case "restore":
-                    await Restore.Run(docset, options, log);
+                    await Restore.Run(source, options, log);
                     break;
                 case "build":
-                    await Build.Run(docset, options, log);
+                    await Build.Run(source, options, log);
                     break;
             }
         }
 
-        private static (string command, string docset, CommandLineOptions options) ParseCommandLineOptions(string[] args)
+        private static (string command, string source, CommandLineOptions options) ParseCommandLineOptions(string[] args)
         {
-            return (args[0], args[1], default);
+            var command = string.Empty;
+            var source = ".";
+            var options = new CommandLineOptions();
+
+            ArgumentSyntax.Parse(args, syntax =>
+            {
+                // Restore command
+                syntax.DefineCommand("restore", ref command, "restores dependencies before build");
+                syntax.DefineParameter("source", ref source, "docset path that contains docfx.yml");
+
+                // Build command
+                syntax.DefineCommand("build", ref command, "builds a folder containing docfx.yml");
+                syntax.DefineOption("o|out", ref options.Output, "output folder");
+                syntax.DefineOption("log", ref options.Log, "path to log file");
+                syntax.DefineOption("locale", ref options.BuildLocale, "locale to build");
+                syntax.DefineOptionList("github-token", ref options.GitHubTokens, "GitHub personal access tokens to call github API");
+                syntax.DefineOption("stable", ref options.Stable, "produces stable output for comparison in a diff tool");
+                syntax.DefineParameter("source", ref source, "docset path that contains docfx.yml");
```

Comment:
>source [](start = 40, length = 6)

why it's parsed twice?

Reply:
different command have to define their own option

---
In reply to: [184253393](https://github.com/dotnet/docfx/pull/2690#discussion_r184253393) [](ancestors = 184253393)

Reply:
it should be a common options

---
In reply to: [184261151](https://github.com/dotnet/docfx/pull/2690#discussion_r184261151) [](ancestors = 184261151,184253393)